esp-hal/src/utils/macros.rs

```diff
@@ -0,0 +1,117 @@
+#     loop {}
+# }
+# fn main() {
+#     let mut peripherals = esp_hal::init(esp_hal::Config::default());
+"#
+    };
+}
+
+#[doc(hidden)]
+/// Shorthand to define enums with From implementations.
+#[macro_export]
+macro_rules! any_enum {
+    ($(#[$meta:meta])* $vis:vis enum $name:ident {
+        $(
+            $(#[$variant_meta:meta])*
+            $variant:ident($inner:ty)
+        ),* $(,)?
+    }) => {
+        $(#[$meta])*
+        $vis enum $name {
+            $(
+                $(#[$variant_meta])*
+                $variant($inner),
+            )*
+        }
+
+        $(
+            $(#[$variant_meta])*
+            impl From<$inner> for $name {
+                fn from(inner: $inner) -> Self {
+                    $name::$variant(inner)
+                }
+            }
+        )*
+    };
+}
+
+#[doc(hidden)]
+/// Shorthand to define AnyPeripheral instances.
+#[macro_export]
+macro_rules! any_peripheral {
+    ($(#[$meta:meta])* $vis:vis peripheral $name:ident {
+        $(
+            $(#[cfg($variant_meta:meta)])*
+            $variant:ident($inner:ty)
+        ),* $(,)?
+    }) => {
+        paste::paste! {
+            $(#[$meta])*
+            $vis struct $name([< $name Inner >]);
+            impl $crate::private::Sealed for $name {}
+
+            impl $crate::dma::PeripheralMarker for $name {
+                #[inline(always)]
+                fn peripheral(&self) -> $crate::system::Peripheral {
+                    match &self.0 {
+                        $(
+                            $(#[cfg($variant_meta)])*
+                            [<$name Inner>]::$variant(inner) => inner.peripheral(),
+                        )*
+                    }
+                }
+            }
+
+            impl $crate::peripheral::Peripheral for $name {
+                type P = $name;
+
+                unsafe fn clone_unchecked(&self) -> Self::P {
+                    match &self.0 {
+                        $(
+                            $(#[cfg($variant_meta)])*
+                            [<$name Inner>]::$variant(inner) => $name::from(inner.clone_unchecked()),
+                        )*
+                    }
+                }
+            }
+
+            enum [< $name Inner >] {
+                $(
+                    $(#[cfg($variant_meta)])*
+                    $variant($inner),
+                )*
+            }
+
+            $(
+                $(#[cfg($variant_meta)])*
+                impl From<$inner> for $name {
+                    fn from(inner: $inner) -> Self {
+                        Self([< $name Inner >]::$variant(inner))
+                    }
+                }
+            )*
+
+            $(
+                $(#[cfg($variant_meta)])*
+                impl $inner {
+                    /// Returns a type-erased version of this peripheral.
+                    pub fn degrade(self) -> $name {
+                        self.into()
+                    }
+                }
```

Comment:
I'm thinking it might be better to do erasure at the driver level rather than the peripheral level.

I can't think of any reason a user would call this directly, considering the `new`/`new_typed` pattern exists.

I'm also thinking it might be better to defer the type erasure until after the user is done constructing the driver.
```Rust
let mut spi = Spi::new(peripherals.SPI2, 100.kHz(), SpiMode::Mode0)
    .with_pins(sclk, mosi, miso, cs)
    // .degrade() // Type erasure happens after SPI pins have been assigned. i.e. QPI vs OPI
    .with_dma(dma_channel.configure_for_async(false, DmaPriority::Priority0))
    .degrade() // Type erasure happens after DMA channel compatibility check has happened.
    .with_buffers(dma_rx_buf, dma_tx_buf);
```

Though I have a strong feeling you consider this `.degrade()` call to be boilerplate :smile:.

Reply:
I guess this (your `.degrade()` call) would be a non-issue if we had a separate driver builder. Syntactically, it would mean an additional `.build()` which is indeed not something I consider ideal ideal. That is somewhat of an orthogonal issue, though, and I'd also want to think about other peripherals and general consequences w.r.t. built drivers and changing settings before doing anything with it.

`fn degrade` exists mostly for cargo cult reasons - embassy has it for pins, we have it for pins, we might as well have it for peripherals. It's not necessary, as erased peripherals are all `Into<Any*>`, so I might get rid of this. `Into<Any*>` came into the picture somewhat later.